mxnet/profile.go

```diff
@@ -38,44 +38,78 @@ type Profile struct {
 	filename  string
 }
 
-type ProfileMode int
-
+// profile type
+type ProfileMode string
+// profile options
 const (
-	ProfileSymbolicOperators = ProfileMode(0)
-	ProfileAllOperators      = ProfileMode(1)
+        ProfileAllDisable = ProfileMode(0)
+        ProfileSymbolicOperatorsDisable = ProfileMode(0)
+        ProfileImperativeOperatorsDisable = ProfileMode(0)
+        ProfileMemoryDisable = ProfileMode(0)
+        ProfileApiDisable = ProfileMode(0)
+        ProfileContiguousDumpDisable = ProfileMode(0)
+        ProfileAllEnable  = ProfileMode(1)
+        ProfileSymbolicOperatorsEnable = ProfileMode(1)
+        ProfileImperativeOperatorsEnable = ProfileMode(1)
+        ProfileMemoryEnable = ProfileMode(1)
+        ProfileApiEnable = ProfileMode(1)
+        ProfileContiguousDumpEnable = ProfileMode(1)
+        ProfileDumpPeriod = ProfileMode(1)
 )
 
 // go binding for MXSetProfilerConfig()
-// param mode kOnlySymbolic: 0, kAllOperator: 1
-// param filename output filename
-func NewProfile(mode ProfileMode, tmpDir string) (*Profile, error) {
-	if tmpDir == "" {
-		tmpDir = filepath.Join(config.App.TempDir, "mxnet", "profile")
-	}
-	if !com.IsDir(tmpDir) {
-		os.MkdirAll(tmpDir, os.FileMode(0755))
-	}
-	filename, err := tempFile(tmpDir, "profile-", ".json")
-	if err != nil {
-		return nil, errors.Errorf("cannot create temporary file in %v", tmpDir)
-	}
-
-	cs := C.CString(filename)
-	defer C.free(unsafe.Pointer(cs))
-	success, err := C.MXSetProfilerConfig(C.int(mode), cs)
-	if err != nil {
-		return nil, err
-	}
-	if success != 0 {
-		return nil, GetLastError()
-	}
+// param profile_options map of profiling options
+// param tmpDir output filepath
+// CHANGE: modified interface to support new profiling options
+// TODO: make conversion neater if possible
+// TODO: check functionality
+func NewProfile(profileOptions map[string]ProfileMode, tmpDir string) (*Profile, error) {
+        if tmpDir == "" {
+                tmpDir = filepath.Join(config.App.TempDir, "mxnet", "profile")
+        }
+        if !com.IsDir(tmpDir) {
+                os.MkdirAll(tmpDir, os.FileMode(0755))
+        }
+        filename, err := tempFile(tmpDir, "profile-", ".json")
+        if err != nil {
+                return nil, errors.Errorf("cannot create temporary file in %v", tmpDir)
+        }
+
+        cs := C.CString(filename)
```

Comment:
put `defer C.free(unsafe.Pointer(cs));` here and delete the code at the bottom

Reply:
cool!